tests/platform_tests/counterpoll/test_counterpoll_watermark.py

```diff
@@ -0,0 +1,233 @@
+                                                       CounterpollConstants.PG_WATERMARK_STAT_TYPE]},
+                 CounterpollConstants.PG_DROP: {MAPS: [PG_MAPS],
+                                                CounterpollConstants.TYPE: [CounterpollConstants.PG_DROP_STAT_TYPE]}
+                 }
+
+WATERMARK_COUNTERS_DB_STATS_TYPE = ['USER_WATERMARKS', 'PERSISTENT_WATERMARKS', 'PERIODIC_WATERMARKS']
+
+
+@pytest.fixture(scope='module')
+def dut_vars(duthosts, enum_rand_one_per_hwsku_hostname, request):
+    inv_files = get_inventory_files(request)
+    dut_vars = get_host_visible_vars(inv_files, enum_rand_one_per_hwsku_hostname)
+    yield dut_vars
+
+
+def test_counterpoll_queue_watermark_pg_drop(duthosts, localhost, enum_rand_one_per_hwsku_hostname, dut_vars):
+    """
+    @summary: Verify FLEXCOUNTERS_DB and COUNTERS_DB content after `counterpoll queue/watermark/queue enable`
+
+    this test runs the following steps once with config reload between counterpolls enabled,
+     and once with switch reboot:
+    1. Disable the counter polling for all counters
+    2. save and reboot or reload configuration (randomized)
+    3. run one of the (type[watermark,queue,pg-drop]) (randomized)
+        a. queue: enable the queue counter polling, check whether queue map is generated in COUNTERS_DB,
+           no WATERMARK or PG-DROP stats in FLEX_COUNTER_DB
+        b. watermark: enable the watermark counter polling, check whether PG and queue maps are generated in COUNTER_DB
+           and both PG and QUEUE WATERMARK stats exists in FLEX_COUNTER_DB, without any QUEUE or PG-DROP stats
+           for watermark also check WATERMARK stats in COUNTERS_DB (USER/PERSISTENT/PERIODIC WATERMARKS)
+        c. pg-drop: enable the pg-drop counter polling, check whether PG map is generated in COUNTERS_DB
+           no WATERMARK or QUEUE stats in FLEX_COUNTER_DB
+    4. enables all three counterpolls (queue,watermark,pg-drop) and count stats per type
+    """
+    duthost = duthosts[enum_rand_one_per_hwsku_hostname]
+    skip_release(duthost, ["202205", "202111", "202106", "202012", "201911", "201811", "201803"])
+
+    counted_dict = {}
+    # choosing only one between reload / reboot due to test duration limitations
+    with allure.step("choosing random config apply method"):
+        config_apply_method = random.choice(["config reload", "switch reboot"])
+    with allure.step("disabling all counterpolls"):
+        ConterpollHelper.disable_counterpoll(duthost, CounterpollConstants.COUNTERPOLL_MAPPING.values())
+
+    # verify relevant counterpolls (queue/watermark/pg-drop) are disabled
+    with allure.step("Verifying initial output of {} on {} ..."
+                     .format(CounterpollConstants.COUNTERPOLL_SHOW, duthost.hostname)):
+        verify_all_counterpoll_status(duthost, DISABLE)
+
+    with allure.step("saving config on dut {} after counterpoll disable...".format(duthost.hostname)):
+        duthost.shell('config save -y')
+
```

Comment:
Please consider backup the existing config_db.json before running config save. We have a fixture `backup_and_restore_config_db` to do that

Reply:
isn't it being done automatically when running the test ?
I noticed that the counterpoll configuration was saved and restored when I ran partial test during the development before enabling all counterpolls part at the end of this test.
i.e. when my test only wanted to test one of the counterpoll enabled whlie all of the others disabled, and the switch had all counterpoll enabled, this state of all counterpolls enabled was restored at the end of the test

if needed I can add this fixture usage

Reply:
fixed, added the use of backup_and_restore_config_db fixture:

17:03:52 duthost_utils._backup_and_restore_config L0038 INFO   | Restore /etc/sonic/config_db.json with /etc/sonic/config_db.json.before_test_function